File: 01-expense-tracker/js-version/src/components/Expenses/Expenses.js
import { useState } from 'react'

import Card from '../UI/Card'
import ExpensesFilter from './ExpensesFilter'
import ExpensesList from './ExpensesList'

import './Expenses.css'
import ExpensesChart from './ExpensesChart'

const Expenses = (props) => {
	const [filteredYear, setFilteredYear] = useState('2021')

	const filterChangeHandler = (year) => {
		setFilteredYear(year)
	}

	const filteredExpenses = props.items
		.filter((item) => item.date.getFullYear().toString() === filteredYear)
		.sort((a, b) => b.date.getTime() - a.date.getTime())

	return (
		<div>
			<Card className="expenses">
				<ExpensesFilter
					selected={filteredYear}
					onChangeFilter={filterChangeHandler}
				/>
				<ExpensesChart expenses={filteredExpenses} />

				<ExpensesList items={filteredExpenses} />
			</Card>
		</div>
	)
}

export default Expenses

File: 01-expense-tracker/kotlin-version/src/main/kotlin/components/expenses/Expenses.kt
package components.expenses

import commons.utils.FunctionalComponentDelegate
import commons.utils.css
import components.ui.card
import kotlinx.css.*
import model.Expense
import react.RProps
import react.dom.div
import react.useState

interface ExpensesProps : RProps {

	var items: List<Expense>

}

val expenses by FunctionalComponentDelegate<ExpensesProps> { props ->
	var filteredYear by useState("2021")

	val filterChangeHandler = { year: String ->
		filteredYear = year
	}

	val filteredExpenses = props.items
		.filter { expense -> expense.date.getFullYear().toString() == filteredYear }
		.sortedByDescending { it.date.getTime() }

	div {
		card {
			css {
				padding(1.rem)
				backgroundColor = rgb(31, 31, 31)
				margin(2.rem, LinearDimension.auto)
				width = 50.rem
				maxWidth = 95.pct
			}

			expensesFilter {
				attrs.selected = filteredYear
				attrs.onChangeFilter = filterChangeHandler
			}

			expensesChart {
				attrs.expenses = filteredExpenses
			}

			expensesList {
				attrs.items = filteredExpenses
			}
		}
	}
}
